### PR TITLE
Add impersonation functionality

### LINKED ISSUE #4: Add impersonation feature
We have this in apbears, and it's turned out to be quite useful. It allows superusers to "impersonate" a regular user, i.e. they appear to be logged in as that user so the admin can see exactly what the user sees. This should be built in to this gem.

### PR DESCRIPTION
**Pull request by darinwilson**
This adds the "impersonation" functionality that we have in APBears to this gem, so that it can be available in all of the UCB Rails apps.

This should resolve #4 

### Motivation

Production issues are often data-specific: the system will appear to be working correctly for one user, while another user encounters a bug. To help diagnose these kinds of issues, it's useful for admins to be see exactly what the end users are seeing. Allowing admins to impersonate other users helps to achieve this.

### Implementation

Impersonations are started and stopped by adding and updating records in the database. I didn't want to try to do anything tricky to the session object, just in case some future app used an insecure session storage strategy that made impersonation easy to spoof.

Instead, this PR adds a new `impersonations` table that tracks each time a user wants to impersonate another user. Each record has a column with the id of the admin user, the id of the user they want to impersonate, and an `active` flag. When an `impersonation` record has its `active` flag set to true, then the admin is impersonating another user.

When the admin decides to stop impersonating, the `active` flag is set to false, but the record is saved in the table. This creates a log of who impersonated whom and when, and also allows us to present a "recent impersonations" screen to the admins, making it easy to return to impersonations they use frequently.

### Workflow

This PR adds a new admin screen at `/admin/impersonations`. By default, only superusers can access this screen, but host apps can customize this behavior by overriding the `User#can_impersonate?` method.

On the impersonations screen, the admin can select one of their recent impersonations, or search for a user using an autocomplete-enabled search box.

When an impersonation is initiated, a new `impersonations` record is written to the database, and the logic for determining `current_user` changes.

We still look at the user id stored in the session to determine who the user is, but once we've done that, we also look to see if they have an active impersonation record in the database. If so, `current_user` returns the impersonated user, not the admin user.

If you want to know who the actual logged-in user is, you can use `logged_in_user` instead of `current_user` - this will always return the "real" user, whether or not there's an active impersonation.

For example:

Alice is an admin who has logged into the system:
   * `logged_in_user` returns Alice
   * `current_user` returns Alice

Alice then starts impersonating Bob:
   * `logged_in_user` returns Alice
   * `current_user` returns Bob

Alice stops impersonating Bob:
   * `logged_in_user` returns Alice
   * `current_user` returns Alice

Most of our apps rely on `current_user` to determine what should or should not be displayed to the user, so this seemed to be the best way to make the addition of this feature seamless to those apps.

### Upgrading

This is a substantial new feature, so there a few steps involved to integrate it into an already-existing application. [This wiki page](https://github.com/ucb-ist-eas/ucb_rails_user/wiki/Upgrading-to-version-3.0) describes those steps in detail.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -169,6 +196,8 @@ Then, from within any request or integration spec, you should be able to do this
 
 and the user should now be logged in.
 
+*NOTE:* For system specs, the logic is a little different - use `system_login_user` rather than `login_user` in these specs.
```
Comment (brandondrew) [inline README.md]: Is this supposed to be `logged_in_user`?

File: app/controllers/ucb_rails_user/concerns/controller_methods.rb [hunk 2]
```diff
@@ -11,26 +11,30 @@ module UcbRailsUser::Concerns::ControllerMethods
...
-      logger.debug 'recalc of current_user'
-      user_session_manager.current_user(session[:uid])
+  def logged_in_user
+    @logged_in_user ||= begin
+      logger.debug 'recalc of logged_in_user'
+      user_session_manager&.current_user(session[:uid])
     end
   end
 
+  def current_user
```
Comment (brandondrew) [inline app/controllers/ucb_rails_user/concerns/controller_methods.rb]: This is the heart of it all... very elegant 👍

File: app/controllers/ucb_rails_user/concerns/impersonations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    logged_in_user.stop_impersonation!
+    flash[:info] = "You are no longer impersonating."
+    redirect_to request.referer || "/"
+  end
+
+  private
+
+  def check_permissions
+    unless logged_in_user.can_impersonate?
+      # TODO: I'm not crazy about this - should we provide some generic error pages for common situations?
```
Comment (brandondrew) [inline app/controllers/ucb_rails_user/concerns/impersonations_controller.rb]: this seems perfect for the time being... we can see how important something else is as we use it, and what exactly we want...

File: README.md (hunk 1 above)
Comment (darinwilson) [reply README.md]: No - this is different. It's a method used in request or feature specs that simulates the user logging in.

**Comment by darinwilson** [comment]
> > I didn't want to try to do anything tricky to the session object, just in case some future app used an insecure session storage strategy that made impersonation easy to spoof.
> Is this the same strategy that APBears uses?

It's similar, but not exact. APBears has an `impersonation` table but also a `user_switches` table, which are related, but I honestly don't fully understand how. Also, the impersonation feature is woven into more places in APBears because it has a more complicated role system. I tried to take the core of what APBears was doing and work it into way that would be as seamless as possible for apps that wanted it to "just work", which still providing some hooks for customization when needed.

> Without more specifics, it _seems_ like a good strategy to avoid messing around with the session object, but I'm curious to know if you have a more specific scenario in mind, and (if so) would it make an already existing security hole bigger? Could it create a security hole where none would otherwise exist? 

Since we're keeping everything related to impersonation on the server-side, and within the walled garden of an authenticated user session, I don't think we're making the system any less secure than it might already be. I believe the only way to crack through would be to convince the system that you're a superuser, and at that point, you can do worse things that impersonate another user. That danger already exists (although it would be difficult to do), but I don't believe this feature is making it any worse.

> It seems very good to make `logged_in_user` available as well... but I wonder if we risk a certain scenario where the app developer decides to restrict certain things to only the actual logged-in user, and then an impersonator is unable to reproduce a certain behavior... is there any way to avoid this?

I don't think there's a way to do it without really complicating the feature, which I'd rather not do. I think it's in the developer's best interest not to try do anything sneaky with `logged_in_user` as they're more than likely get a frustrated admin who wants to know why the site is misbehaving.

> Perhaps a prior question should be "when will we need to use `logged_in_user`?" 

The main thing we're using it for right now is rendering the nav. We want to provide a "Stop impersonating" link where we'd normally supply a "Logout" link when an admin is impersonating another user, and to do that, we need to know that `logged_in_user` has an active impersonation. Bu the nav is not actually supplied by this gem. We have a standard one that gets added when an app is generated via [ucb_rails_cli](https://github.com/ucb-ist-eas/ucb_rails_cli) but apps may want to customize it, or replace it entirely.

**Pull request merged**